### ISSUE TITLE
token_fetch fails with 401 error

### ISSUE THREAD

**Issue by ekarsten** [r-lib/gargle#165]
I recently updated from to gargle 1.0.0 from gargle 0.5.0. When I try to get a token, gargle seems to step through all the authentication steps in the browser and generate an authorization code, but then it fails with a 401 error. The output is:

```r
> token <- gargle::credentials_user_oauth2(
+   scopes = "https://www.googleapis.com/auth/spreadsheets",
+   gargle_quiet = FALSE
+ )
httpuv not installed, defaulting to out-of-band authentication
Enter authorization code: ****** Censored for my privacy *******
Error in oauth2.0_access_token(endpoint, app, code = code, user_params = user_params,  : 
  Unauthorized (HTTP 401). Failed to get an access token.
```
My session info is:

```r
> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.0.2 tools_4.0.2
```

**Comment by jennybc** [comment]
I don't have any sort of hunch what this is about. I've done lots of auth in the past day or so *without httpuv* and I'm not having any trouble. Is this reproducible for you?

**Comment by jennybc** [comment]
The generic advice for getting visibility into gargle and its tokens is this:

https://gargle.r-lib.org/articles/troubleshooting.html

**Comment by ekarsten** [comment]
I was able to replicate the same 401 error on a different computer that had never had a prior version of gargle installed before with a different gmail account (that I had not previously authorized any API access). Now, it's possible that some of the issue is caused by a conflict with another package I have installed in both places. I have yet to try on a clean install of R. It doesn't seem like the issue is stemming from bad cached tokens given that it can be replicated on a computer with no cached tokens.

**Comment by jennybc** [comment]
This is a shot in the dark but is there any chance of overlap with #143 or #111?

In any case, some of the troubleshooting techniques from #111 (especially as you scroll down) may eventually become necessary.

**Comment by ekarsten** [comment]
Thank you! This is super helpful. When I run the same thing wrapped in `httr::verbose`, I get the following. Not 100% sure how to parse it, but if I have some spare time in the next couple of days, I'll try to debug some more. When I `install_version` back to 0.5.0, the issue goes away, so that's the workaround I've been using in the meantime.

```
> httr::with_verbose(gargle::token_fetch(
+     scopes = "https://www.googleapis.com/auth/spreadsheets",
+     use_oob = TRUE
+ ), data_in = TRUE, info = TRUE)
trying token_fetch()
trying credentials_service_account()
Error: parse error: premature EOF
                                       
                     (right here) ------^

trying credentials_app_default()
trying credentials_gce()
*  17 bytes stray data read before trying h2 connection
*  Hostname in DNS cache was stale, zapped
*    Trying 169.254.169.254...
*  TCP_NODELAY set
*  Connection timed out after 805 milliseconds
*  Closing connection 3
trying credentials_byo_oauth()
Error: argument "token" is missing, with no default
trying credentials_user_oauth2()
Gargle2.0 initialize
adding 'userinfo.email' scope
loading token from the cache
no matching token in the cache
initiating new token
Enter authorization code: ********************
*  Found bundle for host accounts.google.com: 0x7fa21466eee0 [can multiplex]
*  Connection 1 seems to be dead!
*  Closing connection 1
*    Trying 142.250.68.45...
*  TCP_NODELAY set
*  Connected to accounts.google.com (142.250.68.45) port 443 (#4)
*  ALPN, offering h2
*  ALPN, offering http/1.1
*  successfully set certificate verify locations:
*    CAfile: /etc/ssl/cert.pem
*    CApath: none
*  SSL re-using session ID
*  TLSv1.2 (OUT), TLS handshake, Client hello (1):
*  TLSv1.2 (IN), TLS handshake, Server hello (2):
*  TLSv1.2 (IN), TLS handshake, Certificate (11):
*  TLSv1.2 (IN), TLS handshake, Server key exchange (12):
*  TLSv1.2 (IN), TLS handshake, Server finished (14):
*  TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
*  TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
*  TLSv1.2 (OUT), TLS handshake, Finished (20):
*  TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
*  TLSv1.2 (IN), TLS handshake, Finished (20):
*  old SSL session ID is stale, removing
*  SSL connection using TLSv1.2 / ECDHE-ECDSA-CHACHA20-POLY1305
*  ALPN, server accepted to use h2
*  Server certificate:
*   subject: C=US; ST=California; L=Mountain View; O=Google LLC; CN=accounts.google.com
*   start date: Feb 23 15:43:17 2021 GMT
*   expire date: May 18 15:43:16 2021 GMT
*   subjectAltName: host "accounts.google.com" matched cert's "accounts.google.com"
*   issuer: C=US; O=Google Trust Services; CN=GTS CA 1O1
*   SSL certificate verify ok.
*  Using HTTP2, server supports multi-use
*  Connection state changed (HTTP/2 confirmed)
*  Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
*  Using Stream ID: 1 (easy handle 0x7fa21644fe00)
-> POST /o/oauth2/token HTTP/2
-> Host: accounts.google.com
-> User-Agent: libcurl/7.64.1 r-curl/4.3 httr/1.4.2
-> Accept-Encoding: deflate, gzip
-> Accept: application/json, text/xml, application/xml, */*
-> Content-Type: application/x-www-form-urlencoded
-> Content-Length: 270
-> 
*  Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
>> client_id=***********

*  We are completely uploaded and fine
<- HTTP/2 401 
<- expires: Mon, 01 Jan 1990 00:00:00 GMT
<- pragma: no-cache
<- date: Thu, 11 Mar 2021 19:20:21 GMT
<- cache-control: no-cache, no-store, max-age=0, must-revalidate
<- content-type: application/json; charset=utf-8
<- vary: Origin
<- vary: X-Origin
<- vary: Referer
<- content-encoding: gzip
<- server: scaffolding on HTTPServer2
<- content-length: 81
<- x-xss-protection: 0
<- x-frame-options: SAMEORIGIN
<- x-content-type-options: nosniff
<- 
<<  �

*  Connection #4 to host accounts.google.com left intact
Error: Unauthorized (HTTP 401). Failed to get an access token.
NULL
```

**Comment by jennybc** [comment]
I just saw a 401 (invalid client, unauthorized) and it was ... *exactly that*. I was working with an oauth app I use for development and I literally somehow had the wrong client secret. I think 401s can arise for different reasons and I got more info on mine by using `debug()` on some of the httr functions involved in token acquisition. So I can't say whether you are seeing the same thing as me, but it is worth trying to get more info and making 100% sure that your oauth app is set up correctly.
